flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.util.ExecutorThreadFactory;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.sql.SQLException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class JdbcBatchingOutputFormat<In, JdbcIn, JdbcExec extends JdbcBatchStatementExecutor<JdbcIn>> extends AbstractJdbcOutputFormat<In> {
+	interface RecordExtractor<F, T> extends Function<F, T>, Serializable {
+		static <T> RecordExtractor<T, T> identity() {
+			return x -> x;
+		}
+	}
+
+	interface ExecutorCreator<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
+	}
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(JdbcBatchingOutputFormat.class);
+
+	private final JdbcExecutionOptions executionOptions;
+	private final ExecutorCreator<JdbcExec> statementRunnerCreator;
+	final RecordExtractor<In, JdbcIn> jdbcRecordExtractor;
+
+	private transient JdbcExec jdbcStatementExecutor;
+	private transient int batchCount = 0;
+	private transient volatile boolean closed = false;
+
+	private transient ScheduledExecutorService scheduler;
+	private transient ScheduledFuture<?> scheduledFuture;
+	private transient volatile Exception flushException;
+
+	JdbcBatchingOutputFormat(
+			JdbcConnectionProvider connectionProvider,
+			JdbcExecutionOptions executionOptions,
+			ExecutorCreator<JdbcExec> statementExecutorCreator,
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
+		this.statementRunnerCreator = statementExecutorCreator;
+		this.jdbcRecordExtractor = recordExtractor;
+	}
+
+	/**
+	 * Connects to the target database and initializes the prepared statement.
+	 *
+	 * @param taskNumber The number of the parallel instance.
+	 */
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		super.open(taskNumber, numTasks);
+		jdbcStatementExecutor = statementRunnerCreator.apply(getRuntimeContext());
+		try {
+			jdbcStatementExecutor.open(connection);
+		} catch (SQLException e) {
+			throw new IOException("unable to open JDBC writer", e);
+		}
+		if (executionOptions.getBatchIntervalMs() != 0 && executionOptions.getBatchSize() != 1) {
+			this.scheduler = Executors.newScheduledThreadPool(1, new ExecutorThreadFactory("jdbc-upsert-output-format"));
+			this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(() -> {
+				synchronized (JdbcBatchingOutputFormat.this) {
+					if (!closed) {
```

Comment:
Are we sure about the concurrency here? Here `closed` is guarded by a lock but all the remaining usages are not. Here we are simply reading the value which is `volatile`, but in the `close()` we have a `if (!closed) { closed = true;  ... ` which is not atomic and then we `cancel` the `scheduledFuture` executing this code.

Reply:
`close()` is also `synchronized` so these sections can't execute simultaneously.

`volatile` seems unnecessary here; I didn't remove this modifier because I tried to keep the internal logic untouched where possible (performance-wise it doesn't matter since it's not in the hot path).